### ISSUE TITLE
Feature request: Call parent::__construct before other constructor code

### ISSUE THREAD

**Issue by Levivb** [slevomat/coding-standard#447]
Hey,

I've got a feature request from which it origin stems from the language `c` (or `c++` if you will).

Given class A extends class B. Both class A and class B expect certain arguments in their constructor:
```php
class B {
  /** @var ThingB */
  private $b;
  public function __construct(ThingB $b){
    $this->b = $b;
  }
}

class A extends B {
  /** @var ThingA */
  private $a;
  public function __construct(ThingB $b, ThingA $a){
    // the culprit
  }
}
```

Since class A should pass `$b` to it's parent constructor there are two possible solutions for the constructor of class `A`:
```php
  public function __construct(ThingB $b, ThingA $a){
    $this->a = $a;
    parent::__construct($b);
  }
```
or
```php
  public function __construct(ThingB $b, ThingA $a){
    parent::__construct($b);
    $this->a = $a;
  }
```

In more low level languages, you'r forced to call the parent constructor first before assigning variables to the class properties because of the simple reason of memory allocation. The first solution, variable `$a` is written to memory without having any space allocated first, since that is part of the parent constructors responsibility (or all the way up to the first base class).

For php, not a big deal in which order the properties are assigned and whether the parent constructor is called before or after the assignment of the properties (memory wise). It does, however, display intent.

For the first solution, the `$a` variable is assigned to the `a` property of the class before calling the parent constructor. This gives the impression the parent class is in need of the property assigned in the subclass.

My request: I would like to see a sniff which enforces the calling of `parent::__construct` (if called) before any other code in the constructor.
I can't think of a reason not to do this, and if a class really should call the `parent::__construct` halfway through it's own constructor, it can always be made explicit by ignoring the rule locally.

**Comment by simPod** [comment]
I don't think such rule is something that could be considered coding style. It always depends on current design of things.

Compliance with it breaks following code https://3v4l.org/oP2Tr
```php
<?php

abstract class X {
    protected function __construct() {
        $this->a();
    }
    protected abstract function a();
}

final class A extends X {
    private $prop;
    
    public function __construct() {
        parent::__construct();
        
        $this->prop = 1;
        
        // parent::__construct();
    }
    
    protected function a() : int {
        return $this->prop;
    }
}

new A();
```

> Fatal error: Uncaught TypeError: Return value of A::a() must be of the type integer, null returned

**Comment by Majkl578** [comment]
It can't be part of a CodeSniffer-based CS:
* can't check whether ancestor has a constructor,
* can't check the signature of ancestor's constructor.
